File: scripts/run_inference.mjs
/**
 * MELD-PostOp CLI inference — Node.js
 *
 * Runs the identical inference pipeline as the browser web worker, using the
 * same shared inference.js / preprocess.js / nifti.js modules. The only
 * environment-specific pieces are:
 *   - File I/O: node:fs instead of fetch
 *   - Gzip decompression: node:zlib instead of DecompressionStream
 *   - ONNX runtime: onnxruntime-node instead of ort.js WASM
 *
 * Usage:
 *   # Single file
 *   node scripts/run_inference.mjs --input scan.nii.gz --output mask.nii \
 *     [--model-dir web/models/] [--folds 0] [--threads 4]
 *
 *   # Batch (input directory → output directory)
 *   node scripts/run_inference.mjs --input-dir /data/scans/ --output-dir /data/masks/ \
 *     [--model-dir web/models/] [--folds 0,1,2,3,4] [--threads 4]
 *
 * --folds  Comma-separated fold indices to load (default: 0).
 *          Use 0,1,2,3,4 for the full 5-fold ensemble.
 * --model-dir  Directory containing meld_postop_fold{n}.onnx (and optionally
 *              meld_postop_fold{n}_int8.onnx). Default: web/models/
 */

import * as fs   from 'node:fs';
import * as path from 'node:path';
import * as zlib from 'node:zlib';
import * as ort  from 'onnxruntime-node';

import { slidingWindowInference, TARGET_SPACING, PATCH_SIZE, PATCH_OVERLAP, N_CLASSES }
  from '../web/inference.js';
import { resampleTrilinear, foregroundZScore, logitsToSegmentation }
  from '../web/preprocess.js';
import { parseNifti, encodeNifti }
  from '../web/nifti.js';

// ── CLI argument parsing ───────────────────────────────────────────────────

function parseArgs(argv) {
  const args = {};
  for (let i = 0; i < argv.length; i++) {
    if (argv[i].startsWith('--')) {
      const key = argv[i].slice(2);
      args[key] = argv[i + 1] ?? true;
      i++;
    }
  }
  return args;
}

const args = parseArgs(process.argv.slice(2));

const modelDir  = args['model-dir'] ?? 'web/models/';
const foldList  = (args['folds'] ?? '0').split(',').map(Number);
const numThreads = args['threads'] ? parseInt(args['threads'], 10) : 0; // 0 = ORT default
const inputFile  = args['input']      ?? null;
const outputFile = args['output']     ?? null;
const inputDir   = args['input-dir']  ?? null;
const outputDir  = args['output-dir'] ?? null;

if ((!inputFile && !inputDir) || (inputFile && inputDir)) {
  console.error('Provide exactly one of --input <file> or --input-dir <dir>');
  process.exit(1);
}
if (inputFile && !outputFile) {
  console.error('--input requires --output');
  process.exit(1);
}
if (inputDir && !outputDir) {
  console.error('--input-dir requires --output-dir');
  process.exit(1);
}

// ── Logging helpers ────────────────────────────────────────────────────────

function log(msg) { process.stderr.write(msg + '\n'); }

// ── Model loading ──────────────────────────────────────────────────────────

function resolveModelPath(foldIdx, modelDir) {
  const int8 = path.join(modelDir, `meld_postop_fold${foldIdx}_int8.onnx`);
  const fp32 = path.join(modelDir, `meld_postop_fold${foldIdx}.onnx`);
  if (fs.existsSync(int8)) return { filePath: int8, quantized: true };
  return { filePath: fp32, quantized: false };
}

async function loadSessions(foldIndices, modelDir) {
  const sessions = [];
  for (const foldIdx of foldIndices) {
    const { filePath, quantized } = resolveModelPath(foldIdx, modelDir);
    if (!fs.existsSync(filePath)) {
      throw new Error(`Model file not found: ${filePath}`);
    }
    log(`  Loading fold ${foldIdx} — ${quantized ? 'INT8' : 'FP32'}: ${path.basename(filePath)}`);
    const sessionOptions = {};
    if (numThreads > 0) {
      sessionOptions.intraOpNumThreads = numThreads;
    }
    const sess = await ort.InferenceSession.create(filePath, sessionOptions);
    sessions.push(sess);
  }
  return sessions;
}

// ── NIfTI file I/O ─────────────────────────────────────────────────────────

function readNifti(filePath) {
  let buf = fs.readFileSync(filePath);
  // Decompress .nii.gz (gzip magic bytes: 0x1f 0x8b)
  if (buf[0] === 0x1f && buf[1] === 0x8b) {
    buf = zlib.gunzipSync(buf);
  }
  // Convert Node.js Buffer to ArrayBuffer
  return buf.buffer.slice(buf.byteOffset, buf.byteOffset + buf.byteLength);
}

function writeNifti(filePath, arrayBuffer) {
  fs.writeFileSync(filePath, Buffer.from(arrayBuffer));
}

// ── Single-file inference ──────────────────────────────────────────────────

async function runOne(inputPath, outputPath, sessions) {
  const t0 = Date.now();
  log(`\n[${path.basename(inputPath)}]`);

  // 1. Parse NIfTI
  const ab  = readNifti(inputPath);
  const nii = parseNifti(ab);
  log(`  dims ${nii.origDims.join('×')}, spacing ${nii.pixdim.map(v => v.toFixed(3)).join('×')} mm [z,y,x], dtype ${nii.datatype}`);

  // 2. Resample
  const resampled = resampleTrilinear(nii.data, nii.dims, nii.pixdim, TARGET_SPACING);
  log(`  resampled → ${resampled.dims.join('×')}`);

  // 3. Normalise
  const norm      = foregroundZScore(resampled.data);
  const normalised = norm.data;
  log(`  normalised: mean=${norm.mean.toFixed(1)}, std=${norm.std.toFixed(1)}, clip=[${norm.lo.toFixed(1)}, ${norm.hi.toFixed(1)}]`);

  // 4. Sliding-window inference — ensemble over all loaded folds
  const [D, H, W] = resampled.dims;
  const ensembleSoft = new Float32Array(N_CLASSES * D * H * W);

  for (let f = 0; f < sessions.length; f++) {
    log(`  fold ${f + 1}/${sessions.length} inference…`);
    const sess       = sessions[f];
    const inputName  = sess.inputNames[0];
    const outputName = sess.outputNames[0];

    const runModel = async (patchData, shape) => {
      const tensor = new ort.Tensor('float32', patchData, shape);
      const result = await sess.run({ [inputName]: tensor });
      return result[outputName].data;
    };

    const foldSoft = await slidingWindowInference(
      normalised, resampled.dims, runModel, PATCH_SIZE, PATCH_OVERLAP,
      (frac) => process.stderr.write(`\r    ${Math.round(frac * 100)} %  `),
      f,
      log,
    );
    process.stderr.write('\r                \r');

    for (let i = 0; i < ensembleSoft.length; i++) {
      ensembleSoft[i] += foldSoft[i] / sessions.length;
    }
  }

  // 5+6. Argmax + back-resample
  const maskOrig   = logitsToSegmentation(ensembleSoft, resampled.dims, nii.dims, N_CLASSES);
  const maskVoxels = maskOrig.reduce((s, v) => s + v, 0);
  log(`  mask: ${maskVoxels} positive voxels`);

  // 7. Encode and write
  const outNii = encodeNifti(maskOrig, nii.origDims, nii.header);
  writeNifti(outputPath, outNii);
  log(`  → ${outputPath}  (${((Date.now() - t0) / 1000).toFixed(1)} s)`);
}

// ── Main ───────────────────────────────────────────────────────────────────

async function main() {
  log(`Loading ${foldList.length} fold model(s) from ${modelDir}…`);
  const sessions = await loadSessions(foldList, modelDir);
  log(`Models ready.`);

  if (inputFile) {
    await runOne(inputFile, outputFile, sessions);
  } else {
    // Batch mode: process all .nii and .nii.gz files in inputDir
    fs.mkdirSync(outputDir, { recursive: true });
    const files = fs.readdirSync(inputDir)
      .filter(f => f.endsWith('.nii') || f.endsWith('.nii.gz'))
      .sort();

    if (files.length === 0) {
      log(`No .nii / .nii.gz files found in ${inputDir}`);
      process.exit(1);
    }
    log(`Found ${files.length} file(s).`);

    for (let i = 0; i < files.length; i++) {
      const stem   = files[i].replace(/\.nii(\.gz)?$/, '');
      const inPath = path.join(inputDir, files[i]);
      const outPath = path.join(outputDir, `${stem}_mask.nii`);
      log(`[${i + 1}/${files.length}]`);
      await runOne(inPath, outPath, sessions);
    }
    log(`\nDone. ${files.length} file(s) processed.`);
  }
}

main().catch(e => { console.error(e); process.exit(1); });

File: scripts/run_inference.py
#!/usr/bin/env python3
"""
MELD-PostOp Python ONNX inference — GPU provider test

Mirrors the JS pipeline (run_inference.mjs / inference.js / preprocess.js) exactly.
--provider flag lets you compare CPU vs CoreML execution providers.

Usage:
    conda activate meldpostop_onnx
    # Single file
    python scripts/run_inference.py --input scan.nii.gz --output mask.nii \
        [--model-dir web/models/] [--folds 0] [--provider auto|cpu|coreml]

    # Batch
    python scripts/run_inference.py --input-dir /data/scans/ --output-dir /data/masks/ \
        [--model-dir web/models/] [--folds 0,1,2,3,4] [--provider auto]

--provider  auto   (default) try CoreML mlprogram → CoreML NeuralNetwork → CPU
            cpu    force CPU only
            coreml force CoreML (exits if unavailable or model fails to load)
"""

import argparse
import math
import sys
import time
from pathlib import Path

import numpy as np
import onnxruntime as ort
import SimpleITK as sitk
from skimage.transform import resize as sk_resize

# ── Constants (Dataset001_MELDPostOp_v1.0.0 plans.json, 3d_fullres) ──────────
TARGET_SPACING = [0.9999988079071045, 0.9375, 0.9375]  # mm [sz, sy, sx]
PATCH_SIZE     = [128, 160, 112]                         # [D, H, W]
PATCH_OVERLAP  = 0.5
N_CLASSES      = 2
GAUSS_SIGMA    = 0.125  # nnU-Net default


def log(msg):
    print(msg, file=sys.stderr, flush=True)


# ── Provider selection ─────────────────────────────────────────────────────────

def load_session(model_path, provider_arg):
    """
    Create an ORT InferenceSession, trying CoreML formats in order before CPU.
    Returns (session, provider_label) where provider_label describes what's active.
    """
    available = ort.get_available_providers()
    log(f"  ORT available providers: {available}")

    if provider_arg == 'cpu':
        sess = ort.InferenceSession(str(model_path), providers=['CPUExecutionProvider'])
        log(f"  Active providers: {sess.get_providers()}")
        return sess, 'cpu'

    has_coreml = 'CoreMLExecutionProvider' in available
    if not has_coreml:
        if provider_arg == 'coreml':
            log("ERROR: CoreMLExecutionProvider not in this ORT build.")
            sys.exit(1)
        log("  CoreML not available — using CPU.")
        sess = ort.InferenceSession(str(model_path), providers=['CPUExecutionProvider'])
        return sess, 'cpu'

    # Try CoreML ML Program format — supports more 3D ops on M-series (54/77 nodes)
    # but crashes if the model has rank-5 InstanceNormalization (nnU-Net 3d_fullres does)
    try:
        providers = [('CoreMLExecutionProvider', {'ModelFormat': 'MLProgram'}), 'CPUExecutionProvider']
        sess = ort.InferenceSession(str(model_path), providers=providers)
        log(f"  CoreML (MLProgram) loaded. Active: {sess.get_providers()}")
        return sess, 'coreml-mlprogram'
    except Exception as e:
        log(f"  CoreML MLProgram failed: {e}")
        if provider_arg == 'coreml':
            log("ERROR: CoreML MLProgram failed and --provider coreml was requested.")
            sys.exit(1)

    # Try CoreML NeuralNetwork format (legacy — 22/77 nodes on CoreML, rest on CPU)
    try:
        providers = [('CoreMLExecutionProvider', {'ModelFormat': 'NeuralNetwork'}), 'CPUExecutionProvider']
        sess = ort.InferenceSession(str(model_path), providers=providers)
        log(f"  CoreML (NeuralNetwork) loaded. Active: {sess.get_providers()}")
        return sess, 'coreml-neuralnetwork'
    except Exception as e:
        log(f"  CoreML NeuralNetwork failed: {e}")

    # Pure CPU fallback
    log("  Falling back to CPU only.")
    sess = ort.InferenceSession(str(model_path), providers=['CPUExecutionProvider'])
    log(f"  Active providers: {sess.get_providers()}")
    return sess, 'cpu'


# ── Preprocessing (mirrors preprocess.js) ─────────────────────────────────────

def read_nifti(path):
    img = sitk.ReadImage(str(path))
    # GetArrayFromImage returns (z, y, x) order matching SimpleITK / nnU-Net convention
    arr = sitk.GetArrayFromImage(img).astype(np.float32)
    sp_xyz = img.GetSpacing()
    spacing_zyx = [sp_xyz[2], sp_xyz[1], sp_xyz[0]]
    return arr, spacing_zyx, img


def resample_trilinear(arr, src_spacing, tgt_spacing):
    dst_dims = [
        round(arr.shape[i] * src_spacing[i] / tgt_spacing[i])
        for i in range(3)
    ]
    return sk_resize(
        arr, dst_dims,
        order=1, mode='edge', anti_aliasing=False, preserve_range=True,
    ).astype(np.float32)


def foreground_zscore(arr):
    fg = np.sort(arr[arr != 0])
    lo = fg[int(0.005 * len(fg))]
    hi = fg[int(0.995 * len(fg))]
    clipped_fg = np.clip(fg, lo, hi)
    mean = float(clipped_fg.mean())
    std  = float(clipped_fg.std()) + 1e-8
    return ((np.clip(arr, lo, hi) - mean) / std).astype(np.float32), mean, std, lo, hi


# ── Sliding-window inference (mirrors inference.js) ───────────────────────────

def build_gaussian_map(patch_size):
    pD, pH, pW = patch_size
    z = np.arange(pD, dtype=np.float32)
    y = np.arange(pH, dtype=np.float32)
    x = np.arange(pW, dtype=np.float32)
    zz, yy, xx = np.meshgrid(z, y, x, indexing='ij')
    dz = (zz - pD / 2) / (pD * GAUSS_SIGMA)
    dy = (yy - pH / 2) / (pH * GAUSS_SIGMA)
    dx = (xx - pW / 2) / (pW * GAUSS_SIGMA)
    return np.exp(-0.5 * (dz**2 + dy**2 + dx**2)).astype(np.float32)


def compute_sliding_steps(img_dims, patch_dims, step_size):
    steps = []
    for img_d, p_d in zip(img_dims, patch_dims):
        if img_d <= p_d:
            steps.append([0])
            continue
        target_step = p_d * step_size
        num_steps = math.ceil((img_d - p_d) / target_step) + 1
        actual_step = (img_d - p_d) / (num_steps - 1)
        steps.append([round(actual_step * k) for k in range(num_steps)])
    return steps


def sliding_window_inference(data, sess, patch_size, overlap):
    """
    data:  float32 [D, H, W], normalised
    Returns: float32 [N_CLASSES, D, H, W] softmax probabilities
    """
    D, H, W = data.shape
    pD, pH, pW = patch_size
    gauss_map = build_gaussian_map(patch_size)        # [pD, pH, pW]

    accum_soft   = np.zeros((N_CLASSES, D, H, W), dtype=np.float32)
    accum_weight = np.zeros((D, H, W), dtype=np.float32)

    steps_z, steps_y, steps_x = compute_sliding_steps([D, H, W], patch_size, overlap)
    origins = [(z0, y0, x0) for z0 in steps_z for y0 in steps_y for x0 in steps_x]
    n_patches = len(origins)
    log(f"    {len(steps_z)}×{len(steps_y)}×{len(steps_x)} = {n_patches} patches")

    input_name  = sess.get_inputs()[0].name
    output_name = sess.get_outputs()[0].name
    patch_times = []

    for i, (z0, y0, x0) in enumerate(origins):
        # In-bounds slice in global volume
        gz0 = max(0, z0);  gz1 = min(D, z0 + pD)
        gy0 = max(0, y0);  gy1 = min(H, y0 + pH)
        gx0 = max(0, x0);  gx1 = min(W, x0 + pW)
        # Corresponding slice in patch coords
        lz0 = gz0 - z0;  lz1 = gz1 - z0
        ly0 = gy0 - y0;  ly1 = gy1 - y0
        lx0 = gx0 - x0;  lx1 = gx1 - x0

        patch = np.zeros((pD, pH, pW), dtype=np.float32)
        patch[lz0:lz1, ly0:ly1, lx0:lx1] = data[gz0:gz1, gy0:gy1, gx0:gx1]

        t = time.perf_counter()
        raw = sess.run([output_name], {input_name: patch[np.newaxis, np.newaxis]})[0]
        patch_ms = (time.perf_counter() - t) * 1000
        patch_times.append(patch_ms)

        logits = raw[0]  # [2, pD, pH, pW]

        if i == 0:
            log(f"    Patch 0 logits — bg: [{logits[0].min():.2f}, {logits[0].max():.2f}]  "
                f"cavity: [{logits[1].min():.2f}, {logits[1].max():.2f}]  ({patch_ms:.0f} ms)")

        # Numerically stable softmax over class axis
        max_l = logits.max(axis=0, keepdims=True)
        exp_l = np.exp(logits - max_l)
        soft  = exp_l / exp_l.sum(axis=0, keepdims=True)  # [2, pD, pH, pW]

        # Gaussian-weighted accumulation (in-bounds region only)
        w = gauss_map[lz0:lz1, ly0:ly1, lx0:lx1]
        accum_soft[:, gz0:gz1, gy0:gy1, gx0:gx1] += soft[:, lz0:lz1, ly0:ly1, lx0:lx1] * w
        accum_weight[gz0:gz1, gy0:gy1, gx0:gx1]  += w

        if i % max(1, n_patches // 20) == 0 or i == n_patches - 1:
            pct = round((i + 1) / n_patches * 100)
            print(f"\r    {pct:3d} %  ({patch_ms:.0f} ms/patch)", end='', file=sys.stderr, flush=True)

    print('', file=sys.stderr)
    avg_ms = sum(patch_times) / len(patch_times)
    log(f"    avg {avg_ms:.0f} ms/patch over {n_patches} patches")

    accum_weight = np.maximum(accum_weight, 1e-8)
    return accum_soft / accum_weight[np.newaxis]  # [N_CLASSES, D, H, W]


# ── Postprocessing ─────────────────────────────────────────────────────────────

def logits_to_segmentation(soft_vol, orig_dims):
    mask_resampled = np.argmax(soft_vol, axis=0).astype(np.float32)  # [D, H, W]
    return sk_resize(
        mask_resampled, orig_dims,
        order=0, mode='edge', anti_aliasing=False, preserve_range=True,
    ).astype(np.uint8)


# ── File I/O ───────────────────────────────────────────────────────────────────

def write_nifti(mask, ref_img, output_path):
    out_img = sitk.GetImageFromArray(mask)
    out_img.CopyInformation(ref_img)
    sitk.WriteImage(out_img, str(output_path))


# ── Model loading ──────────────────────────────────────────────────────────────

def resolve_model_path(fold_idx, model_dir):
    int8 = Path(model_dir) / f'meld_postop_fold{fold_idx}_int8.onnx'
    fp32 = Path(model_dir) / f'meld_postop_fold{fold_idx}.onnx'
    if int8.exists():
        return int8, True
    return fp32, False


# ── Single-file inference ──────────────────────────────────────────────────────

def run_one(input_path, output_path, sessions):
    t0 = time.perf_counter()
    log(f"\n[{Path(input_path).name}]")

    arr, spacing_zyx, ref_img = read_nifti(input_path)
    orig_dims = arr.shape
    log(f"  dims {list(arr.shape)}, spacing {[f'{s:.3f}' for s in spacing_zyx]} mm [z,y,x]")

    resampled = resample_trilinear(arr, spacing_zyx, TARGET_SPACING)
    log(f"  resampled → {list(resampled.shape)}")

    normalised, mean, std, lo, hi = foreground_zscore(resampled)
    log(f"  normalised: mean={mean:.1f}, std={std:.1f}, clip=[{lo:.1f}, {hi:.1f}]")

    D, H, W = resampled.shape
    ensemble_soft = np.zeros((N_CLASSES, D, H, W), dtype=np.float32)

    for f_idx, (sess, label) in enumerate(sessions):
        log(f"  fold {f_idx + 1}/{len(sessions)} inference ({label})…")
        fold_soft = sliding_window_inference(normalised, sess, PATCH_SIZE, PATCH_OVERLAP)
        ensemble_soft += fold_soft / len(sessions)

    mask_orig = logits_to_segmentation(ensemble_soft, orig_dims)
    log(f"  mask: {int(mask_orig.sum())} positive voxels")

    write_nifti(mask_orig, ref_img, output_path)
    log(f"  → {output_path}  ({time.perf_counter() - t0:.1f} s)")


# ── Main ───────────────────────────────────────────────────────────────────────

def parse_args():
    p = argparse.ArgumentParser(description='MELD-PostOp ONNX inference (Python)')
    g = p.add_mutually_exclusive_group(required=True)
    g.add_argument('--input',     help='Input NIfTI file')
    g.add_argument('--input-dir', help='Input directory (batch)')
    p.add_argument('--output',     help='Output mask file (with --input)')
    p.add_argument('--output-dir', help='Output directory (with --input-dir)')
    p.add_argument('--model-dir', default='web/models/')
    p.add_argument('--folds', default='0', help='Comma-separated fold indices')
    p.add_argument('--provider', default='auto', choices=['auto', 'cpu', 'coreml'])
    return p.parse_args()


def main():
    args = parse_args()

    if args.input and not args.output:
        sys.exit("ERROR: --input requires --output")
    if args.input_dir and not args.output_dir:
        sys.exit("ERROR: --input-dir requires --output-dir")

    fold_list = [int(f) for f in args.folds.split(',')]

    log(f"Loading {len(fold_list)} fold model(s) from {args.model_dir}…")
    sessions = []
    for fold_idx in fold_list:
        model_path, quantized = resolve_model_path(fold_idx, args.model_dir)
        if not model_path.exists():
            sys.exit(f"ERROR: model not found: {model_path}")
        label = 'INT8' if quantized else 'FP32'
        log(f"  Loading fold {fold_idx} — {label}: {model_path.name}")
        sess, provider_label = load_session(model_path, args.provider)
        sessions.append((sess, provider_label))
    log("Models ready.")

    if args.input:
        run_one(args.input, args.output, sessions)
    else:
        out_dir = Path(args.output_dir)
        out_dir.mkdir(parents=True, exist_ok=True)
        files = sorted(
            f for f in Path(args.input_dir).iterdir()
            if f.suffix in ('.nii',) or f.name.endswith('.nii.gz')
        )
        if not files:
            sys.exit(f"No .nii / .nii.gz files found in {args.input_dir}")
        log(f"Found {len(files)} file(s).")
        for i, in_path in enumerate(files):
            stem = in_path.name.replace('.nii.gz', '').replace('.nii', '')
            log(f"[{i+1}/{len(files)}]")
            run_one(str(in_path), str(out_dir / f"{stem}_mask.nii"), sessions)
        log(f"\nDone. {len(files)} file(s) processed.")


if __name__ == '__main__':
    main()
